### ISSUE TITLE
[Bug] solve_ivp() exceeds t_bound due to rounding error in OdeSolver.step()

### ISSUE THREAD

**Issue by homocomputeris** [scipy/scipy#8672]
In short: in stepsize is something like 1/3, solve_ivp() makes one unnecessary step.

### Reproducing code example:
Let's say we make an ingtegrator that requires fixed time step, e.g. for simplicity Strang:
``` python
class Strang(spi.OdeSolver):
    """Strang method for time-independent case"""

    def __init__(self, sys, t0, y0, t_bound, n_steps,
                 vectorized=False, support_complex=True,
                 **extraneous):
        super().__init__(sys, t0, y0, t_bound,
                         vectorized, support_complex)

        self.h = (t_bound - t0) / n_steps

        self.A, self.B = sys

    def _step_impl(self):
        z = (la.expm(self.h / 2 * self.A())
             @ la.expm(self.h * self.B())
             @ la.expm(self.h / 2 * self.A())
             @ self.y)
        self.t += self.h
        self.y = z
        return True, None
```

Now let's solve a harmonic oscillator. Again, for simplicity, only 10 steps:
``` python
def A():
    a = np.array([[0, 1],
                  [0, 0]])
    return a
def B():
    b = np.array([[0, 0],
                  [-1, 0]])
    return b

t0 = 0.0
tf = 3
t_span = (t0, tf)
u0 = np.ones((2,))
sol = spi.solve_ivp((A, B), t_span, u0, method=Strang, n_steps=10)
print(sol.y.shape)
```

### Error message:
Now let's check how many steps have been made:
``` python
In[2]: sol.t
Out[2]: array([0. , 0.3, 0.6, 0.9, 1.2, 1.5, 1.8, 2.1, 2.4, 2.7, 3. , 3.3])
In[3]: sol.t[-2]
Out[3]: 2.9999999999999996
```
The final time is exceeded, and the last vector in `sol.y` is not the expected solution.

### Comments
I think the same may happen with adaptive steps if solver thinks that last timepoint isn't close enough to `t_bound`.

Obviously, in the example, the penultimate moment in `sol.t` should be the one where the solver finishes.
The stopping criterion should check if a moment is close to `t_bound`and if the next stepsize will be small enough not to exceed `t_bound`. Something like this, maybe:
``` python
class OdeSolver:
# ...
    def step(self):
        if self.status != 'running':
            raise RuntimeError("Attempt to step on a failed or finished "
                               "solver.")

        if self.n == 0 or self.t == self.t_bound:
            # Handle corner cases of empty solver or no integration.
            self.t_old = self.t
            self.t = self.t_bound
            message = None
            self.status = 'finished'
        else:
            t = self.t
            success, message = self._step_impl()

            if not success:
                self.status = 'failed'
            else:
                self.t_old = t
                left_cond = np.allclose(self.t, self.t_bound)
                right_cond = np.allclose(self.t_bound, self.t+self.step_size)
                # if the last step time t is closer to t_bound
                # than the woulb-be step time (t+step_size),
                # we can assume than we have rounding error here
                if self.direction*int(left_cond and not right_cond) != 0:
                    self.status = 'finished'
        return message
```

### Scipy/Numpy/Python version information:
```
1.0.0 1.14.2 sys.version_info(major=3, minor=6, micro=4, releaselevel='final', serial=0)
```
Build with MKL and GCC.

**Comment by drhagen** [comment]
I am having difficulty accepting that the current behavior is wrong. I can imagine someone else coming to report an issue "[BUG] ODE solvers don't integrate to tF if step lands close enough", and I think I would more agree with that being a bug. If I set `tF=10.0`, I would expect the solution to be defined there and not extrapolated.

It is the responsibility of the solver to set `self.t`. In your case, you could change the line `self.t += self.h` to check for being close to `tF` and setting `self.t` to `tF` when it got close. Or even better, check that `n_steps` had been taken and set `self.t` to `tF` because that the implicit guarantee of the constructor.

**Comment by homocomputeris** [comment]
I see your point.
However, how can we be sure that for a solver, that inherits `OdeSolver`,                
```python
if self.direction * (self.t - self.t_bound) >= 0:
    self.status = 'finished'
```
is a correct stopping condition? 
Should a child solver check that `self.h` is such that `self.t` during the last step is no less than `t_bound`?
If so, why does the parent implement it, and doesn't declare this check as an abstract method? If not, and the difference is something like `-10^-40`, the integrator won't stop either. Moreover, if a solver uses complex time, this check doesn't have any sense at all.
To sum up, IMHO, current check seems neither robust nor universal.

**Comment by drhagen** [comment]
> Should a child solver check that `self.h` is such that `self.t` during the last step is no less than `t_bound`?

Yes, during the last step, the `self.t` must be equal to `t_bound` by definition of this being the last step. Solvers must not evaluate the ODE function beyond `t_bound` or provide a solution beyond `t_bound` because the RHS may not be defined there. (This should be explicitly listed in guideline (3), which it currently is not.) The check in `step` should really be `if self.t == self.t_bound`, but for safety, we also check if the solver stepped beyond `t_bound` in violation of the contract.

> If so, why does the parent implement it, and doesn't declare this check as an abstract method?

I assume by "it" you mean the check for the integration being finished. I simply cannot think of any other correct implementation. Stepping beyond `t_bound` is illegal; stopping before `t_bound` is a crash.

> Moreover, if a solver uses complex time, this check doesn't have any sense at all.

I guess complex time is not supported. I have never heard of such a thing.

**Comment by homocomputeris** [comment]
> I assume by "it" you mean the check for the integration being finished. I simply cannot think of any other correct implementation. Stepping beyond t_bound is illegal; stopping before t_bound is a crash.

By 'it' I mean the stopping criterion `if self.direction * (self.t - self.t_bound) >= 0:` implemented in `OdeSolver.step()` that evokes child's `_step_impl()`. AFAIK, current `integrate`'s architecture supposes implementing the latter method and inheriting `step()` without overriding.
`OdeSolver` may declare something like `stop_solver(): NotImplemented` that is delegated to children, not shared by all the solvers, as it is now.

**Comment by drhagen** [comment]
I think I understand what you are asking for. I just don't think there is any other implementation that would be correct in the larger context. `solve_ivp` is expecting the solution to end at `t_bound`. `_step_impl` should simply ensure that this happens. For your Strang example, I would recommend using:

```
self.i_step += 1
self.t = t_bound * (self.i_step / self.n_steps)
```
